File: rust/agama-server/src/web/docs/storage.rs
// Copyright (c) [2024] SUSE LLC
//
// All Rights Reserved.
//
// This program is free software; you can redistribute it and/or modify it
// under the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 of the License, or (at your option)
// any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
// more details.
//
// You should have received a copy of the GNU General Public License along
// with this program; if not, contact SUSE LLC.
//
// To contact SUSE LLC about this file by physical or electronic mail, you may
// find current contact information at www.suse.com.

use utoipa::openapi::{Components, ComponentsBuilder, OpenApi, Paths, PathsBuilder};

use super::{
    common::{IssuesApiDocBuilder, ProgressApiDocBuilder, ServiceStatusApiDocBuilder},
    ApiDocBuilder,
};

pub struct StorageApiDocBuilder;

impl ApiDocBuilder for StorageApiDocBuilder {
    fn title(&self) -> String {
        "Storage HTTP API".to_string()
    }

    fn paths(&self) -> Paths {
        PathsBuilder::new()
            .path_from::<crate::storage::web::__path_actions>()
            .path_from::<crate::storage::web::__path_devices_dirty>()
            .path_from::<crate::storage::web::__path_get_proposal_settings>()
            .path_from::<crate::storage::web::__path_probe>()
            .path_from::<crate::storage::web::__path_product_params>()
            .path_from::<crate::storage::web::__path_set_proposal_settings>()
            .path_from::<crate::storage::web::__path_staging_devices>()
            .path_from::<crate::storage::web::__path_system_devices>()
            .path_from::<crate::storage::web::__path_available_drives>()
            .path_from::<crate::storage::web::__path_candidate_drives>()
            .path_from::<crate::storage::web::__path_available_md_raids>()
            .path_from::<crate::storage::web::__path_candidate_md_raids>()
            .path_from::<crate::storage::web::__path_volume_for>()
            .path_from::<crate::storage::web::dasd::__path_devices>()
            .path_from::<crate::storage::web::dasd::__path_disable>()
            .path_from::<crate::storage::web::dasd::__path_enable>()
            .path_from::<crate::storage::web::dasd::__path_format>()
            .path_from::<crate::storage::web::dasd::__path_probe>()
            .path_from::<crate::storage::web::dasd::__path_set_diag>()
            .path_from::<crate::storage::web::dasd::__path_supported>()
            .path_from::<crate::storage::web::dasd::__path_get_config>()
            .path_from::<crate::storage::web::dasd::__path_set_config>()
            .path_from::<crate::storage::web::iscsi::__path_delete_node>()
            .path_from::<crate::storage::web::iscsi::__path_discover>()
            .path_from::<crate::storage::web::iscsi::__path_initiator>()
            .path_from::<crate::storage::web::iscsi::__path_login_node>()
            .path_from::<crate::storage::web::iscsi::__path_logout_node>()
            .path_from::<crate::storage::web::iscsi::__path_nodes>()
            .path_from::<crate::storage::web::iscsi::__path_update_initiator>()
            .path_from::<crate::storage::web::iscsi::__path_update_node>()
            .path_from::<crate::storage::web::iscsi::__path_set_config>()
            .path_from::<crate::storage::web::zfcp::__path_activate_controller>()
            .path_from::<crate::storage::web::zfcp::__path_activate_controller>()
            .path_from::<crate::storage::web::zfcp::__path_activate_disk>()
            .path_from::<crate::storage::web::zfcp::__path_controllers>()
            .path_from::<crate::storage::web::zfcp::__path_deactivate_disk>()
            .path_from::<crate::storage::web::zfcp::__path_get_global_config>()
            .path_from::<crate::storage::web::zfcp::__path_get_disks>()
            .path_from::<crate::storage::web::zfcp::__path_get_luns>()
            .path_from::<crate::storage::web::zfcp::__path_get_wwpns>()
            .path_from::<crate::storage::web::zfcp::__path_probe>()
            .path_from::<crate::storage::web::zfcp::__path_supported>()
            .build()
    }

    fn components(&self) -> Components {
        ComponentsBuilder::new()
            .schema_from::<agama_lib::issue::Issue>()
            .schema_from::<agama_lib::storage::client::iscsi::ISCSIAuth>()
            .schema_from::<agama_lib::storage::client::iscsi::ISCSIInitiator>()
            .schema_from::<agama_lib::storage::client::iscsi::ISCSINode>()
            .schema_from::<agama_lib::storage::client::iscsi::LoginResult>()
            .schema_from::<agama_lib::storage::model::Action>()
            .schema_from::<agama_lib::storage::model::BlockDevice>()
            .schema_from::<agama_lib::storage::model::Component>()
            .schema_from::<agama_lib::storage::model::Device>()
            .schema_from::<agama_lib::storage::model::DeviceInfo>()
            .schema_from::<agama_lib::storage::model::DeviceSid>()
            .schema_from::<agama_lib::storage::model::DeviceSize>()
            .schema_from::<agama_lib::storage::model::Drive>()
            .schema_from::<agama_lib::storage::model::DriveInfo>()
            .schema_from::<agama_lib::storage::model::Filesystem>()
            .schema_from::<agama_lib::storage::model::LvmLv>()
            .schema_from::<agama_lib::storage::model::LvmVg>()
            .schema_from::<agama_lib::storage::model::Md>()
            .schema_from::<agama_lib::storage::model::Multipath>()
            .schema_from::<agama_lib::storage::model::Partition>()
            .schema_from::<agama_lib::storage::model::PartitionTable>()
            .schema_from::<agama_lib::storage::model::ProposalSettings>()
            .schema_from::<agama_lib::storage::model::ProposalSettingsPatch>()
            .schema_from::<agama_lib::storage::model::ProposalTarget>()
            .schema_from::<agama_lib::storage::model::Raid>()
            .schema_from::<agama_lib::storage::model::ShrinkingInfo>()
            .schema_from::<agama_lib::storage::model::SpaceAction>()
            .schema_from::<agama_lib::storage::model::SpaceActionSettings>()
            .schema_from::<agama_lib::storage::model::UnusedSlot>()
            .schema_from::<agama_lib::storage::model::Volume>()
            .schema_from::<agama_lib::storage::model::VolumeOutline>()
            .schema_from::<agama_lib::storage::model::VolumeTarget>()
            .schema_from::<agama_lib::storage::model::dasd::DASDDevice>()
            .schema_from::<agama_lib::storage::settings::dasd::DASDConfig>()
            .schema_from::<agama_lib::storage::settings::dasd::DASDDeviceConfig>()
            .schema_from::<agama_lib::storage::settings::dasd::DASDDeviceState>()
            .schema_from::<agama_lib::storage::model::zfcp::ZFCPController>()
            .schema_from::<agama_lib::storage::model::zfcp::ZFCPDisk>()
            .schema_from::<crate::storage::web::ProductParams>()
            .schema_from::<crate::storage::web::iscsi::DiscoverParams>()
            .schema_from::<crate::storage::web::iscsi::InitiatorParams>()
            .schema_from::<crate::storage::web::iscsi::LoginParams>()
            .schema_from::<crate::storage::web::iscsi::NodeParams>()
            .schema_from::<crate::storage::web::zfcp::ZFCPGlobalConfig>()
            .build()
    }

    fn nested(&self) -> Option<OpenApi> {
        let mut issues = IssuesApiDocBuilder::new()
            .add(
                "/api/storage/issues",
                "List of storage-related issues",
                "storage_issues",
            )
            .build();
        let status = ServiceStatusApiDocBuilder::new("/api/storage/status").build();
        let progress = ProgressApiDocBuilder::new("/api/storage/progress").build();
        issues.merge(status);
        issues.merge(progress);
        Some(issues)
    }
}

File: web/src/hooks/api/system/storage.ts
/*
 * Copyright (c) [2025] SUSE LLC
 *
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, contact SUSE LLC.
 *
 * To contact SUSE LLC about this file by physical or electronic mail, you may
 * find current contact information at www.suse.com.
 */

import { useCallback } from "react";
import { useSuspenseQuery } from "@tanstack/react-query";
import { systemQuery } from "~/hooks/api/system";
import { flatDevices, findDevices, findDeviceByName } from "~/api/system/storage";
import type { System, storage } from "~/api/system";

const selectSystem = (data: System | null): storage.System => data?.storage;

function useSystem(): storage.System | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectSystem,
  });
  return data;
}

const selectEncryptionMethods = (data: System | null): storage.EncryptionMethod[] =>
  data?.storage?.encryptionMethods || [];

function useEncryptionMethods(): storage.EncryptionMethod[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectEncryptionMethods,
  });
  return data;
}

const enum DeviceGroup {
  AvailableDrives = "availableDrives",
  CandidateDrives = "candidateDrives",
  AvailableMdRaids = "availableMdRaids",
  CandidateMdRaids = "candidateMdRaids",
}

function selectDeviceGroups(data: System | null, groups: DeviceGroup[]): storage.Device[] {
  if (!data?.storage) return [];
  const sids = groups.flatMap((g) => data.storage[g]);
  return findDevices(data.storage, sids);
}

const selectAvailableDrives = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableDrives]);

/**
 * Hook that returns the list of available drives for installation.
 */
function useAvailableDrives(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableDrives,
  });
  return data;
}

const selectCandidateDrives = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateDrives]);

/**
 * Hook that returns the list of candidate drives for installation.
 */
function useCandidateDrives(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateDrives,
  });
  return data;
}

const selectAvailableMdRaids = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableMdRaids]);

/**
 * Hook that returns the list of available MD RAIDs for installation.
 */
function useAvailableMdRaids(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableMdRaids,
  });
  return data;
}

const selectCandidateMdRaids = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateMdRaids]);

/**
 * Hook that returns the list of available MD RAIDs for installation.
 */
function useCandidateMdRaids(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateMdRaids,
  });
  return data;
}

const selectAvailableDevices = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.AvailableDrives, DeviceGroup.AvailableMdRaids]);

/**
 * Hook that returns the list of available devices for installation.
 */
function useAvailableDevices(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectAvailableDevices,
  });
  return data;
}

const selectCandidateDevices = (data: System | null): storage.Device[] =>
  selectDeviceGroups(data, [DeviceGroup.CandidateDrives, DeviceGroup.CandidateMdRaids]);

/**
 * Hook that returns the list of candidate devices for installation.
 */
function useCandidateDevices(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectCandidateDevices,
  });
  return data;
}

const selectDevices = (data: System | null): storage.Device[] => data?.storage?.devices || [];

function useDevices(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectDevices,
  });
  return data;
}

const selectFlattenDevices = (data: System | null): storage.Device[] =>
  data?.storage ? flatDevices(data.storage) : [];

function useFlattenDevices(): storage.Device[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectFlattenDevices,
  });
  return data;
}

function useDevice(name: string): storage.Device | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: useCallback(
      (data: System | null): storage.Device | null => {
        return data?.storage ? findDeviceByName(data.storage, name) : null;
      },
      [name],
    ),
  });
  return data;
}

const selectVolumeTemplates = (data: System | null): storage.Volume[] =>
  data?.storage?.volumeTemplates || [];

function useVolumeTemplates(): storage.Volume[] {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectVolumeTemplates,
  });
  return data;
}

const selectVolumeTemplate = (data: System | null, mountPath: string): storage.Volume | null => {
  const volumes = data?.storage?.volumeTemplates || [];
  return volumes.find((v) => v.mountPath === mountPath);
};

function useVolumeTemplate(mountPath: string): storage.Volume | null {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: useCallback((data) => selectVolumeTemplate(data, mountPath), [mountPath]),
  });
  return data;
}

const selectIssues = (data: System | null): storage.Issue[] => data?.storage?.issues || [];

function useIssues() {
  const { data } = useSuspenseQuery({
    ...systemQuery,
    select: selectIssues,
  });
  return data;
}

export {
  useSystem,
  useEncryptionMethods,
  useAvailableDrives,
  useCandidateDrives,
  useAvailableMdRaids,
  useCandidateMdRaids,
  useAvailableDevices,
  useCandidateDevices,
  useDevices,
  useFlattenDevices,
  useDevice,
  useVolumeTemplates,
  useVolumeTemplate,
  useIssues,
};
